### PR TITLE
Complete doc

### PR DESCRIPTION
**Pull request by arnaud-morvan**
Complete the documentation.

Current pull request generated documentation is available here : https://c2cgeoform.readthedocs.io/en/complete_doc/

### CODE REVIEW

File: docs/developer-guide.rst [hunk 1]
```diff
@@ -1,82 +1,84 @@
...
 
-This page describes how to set up the development environment for
-working on ``c2cgeoform``. It is for developers working on
-``c2cgeoform`` itself, not for developers working on
-``c2cgeoform``-based applications.
+This page describes how to set up the development environment for working on
+c2cgeoform. It is for developers working on c2cgeoform itself, not for
+developers working on c2cgeoform-based applications.
+
+Note that c2cgeoform is a framework and with a
```
Comment (cmangeat) [inline docs/developer-guide.rst]: ???

File: docs/developer-guide.rst [hunk 2]
```diff
@@ -1,82 +1,84 @@
...
 
 .. code-block:: shell
 
-   make test
+   make serve
+
+You can now open the demo project in your favorite browser:
+http://localhost:6543/
+
+Here is it, you're ready to develop in c2cgeoform. Make changes in c2cgeoform,
```
Comment (cmangeat) [inline docs/developer-guide.rst]: Et voilou ?

File: docs/developer-guide.rst [hunk 3]
```diff
@@ -1,7 +1,7 @@
 .. _developer-guide:
 
-Set up development environment
-------------------------------
+Developper guide
```
Comment (cmangeat) [inline docs/developer-guide.rst]: Developper or developer ?

File: docs/developer-guide.rst [hunk 4]
```diff
@@ -82,3 +86,39 @@ http://localhost:6543/
...
+   ssh -A geomapfish-demo.camptocamp.com
+
+   # clone the c2cgeoform repository
+   cd /var/www/vhosts/geomapfish-demo/private
+   git clone git@github.com:camptocamp/c2cgeoform.git
+
+   # generate the c2cgeoform_demo project with mod_wsgi related files
+   APACHE_ENTRY_POINT=c2cgeoform make modwsgi
+
+Create the database as for serving the developement version, see:
```
Comment (cmangeat) [inline docs/developer-guide.rst]: development

File: c2cgeoform/schema.py [hunk 5]
```diff
@@ -40,22 +69,33 @@ def add_unique_validator(self, column, column_id):
...
+        """
         dbsession = self.bindings['dbsession']
         class_ = self.inspector.class_
         return dbsession.query(class_).get(dict_.values())
 
 
 def manytomany_validator(node, cstruct):
     """
-    Validate that cstruct values exists in related table.
+    Validator function that checks ``cstruct`` values exists in related table.
```
Comment (cmangeat) [inline c2cgeoform/schema.py]: check for ?

File: docs/Makefile [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+BUILDDIR      = _build
+
+# Put it first so that "make" without argument is like "make help".
+help:
+	@$(SPHINXBUILD) -M help "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O)
+
+.PHONY: help Makefile
+
+# Catch-all target: route all unknown targets to Sphinx using the new
+# "make mode" option.  $(O) is meant as a shortcut for $(SPHINXOPTS).
```
Comment (julsbreakdown) [inline docs/Makefile]: is a shortcut for $(SPHINXOPTS).

File: docs/conf.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+extensions = [
+    'sphinx.ext.autodoc',
+    'sphinx.ext.githubpages',
+]
+
+# Add any paths that contain templates here, relative to this directory.
+templates_path = ['_templates']
+
+# The suffix(es) of source filenames.
+# You can specify multiple suffix as a list of string:
```
Comment (julsbreakdown) [inline docs/conf.py]: strings

File: docs/conf.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+source_suffix = '.rst'
+
+# The master toctree document.
+master_doc = 'index'
+
+# The language for content autogenerated by Sphinx. Refer to documentation
+# for a list of supported languages.
+#
+# This is also used if you do content translation via gettext catalogs.
+# Usually you set "language" from the command line for these cases.
```
Comment (julsbreakdown) [inline docs/conf.py]: command line in these cases.

File: docs/conf.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+language = None
+
+# List of patterns, relative to source directory, that match files and
+# directories to ignore when looking for source files.
+# This pattern also affects html_static_path and html_extra_path.
```
Comment (julsbreakdown) [inline docs/conf.py]: List of patterns, relative to the source directory, that match files and directories. These will be ignored when looking for source files.
This pattern also affects html_static_path and html_extra_path.

File: docs/conf.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+# -- Options for HTML output -------------------------------------------------
+
+# The theme to use for HTML and HTML Help pages.  See the documentation for
+# a list of builtin themes.
+#
+html_theme = 'sphinx_rtd_theme'
+
+# Theme options are theme-specific and customize the look and feel of a theme
+# further.  For a list of options available for each theme, see the
```
Comment (julsbreakdown) [inline docs/conf.py]: Theme options are theme-specific and enables to further customize the look and feel of a theme
For a list of available options for each theme, see the
documentation.

File: docs/conf.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+# documentation.
+#
+# html_theme_options = {}
+
+# Add any paths that contain custom static files (such as style sheets) here,
+# relative to this directory. They are copied after the builtin static files,
```
Comment (julsbreakdown) [inline docs/conf.py]: Add any paths relative to this directory that contain custom static files (such as style sheets) here,
They are copied after the builtin static files, so a file named "default.css" will overwrite the builtin "default.css".

File: docs/developer-guide.rst (hunk 1 above)
Comment (julsbreakdown) [reply docs/developer-guide.rst]: Note that c2cgeoform is a framework  with a pyramid scaffold....... used to create ....

File: docs/developer-guide.rst [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+Note that c2cgeoform is a framework and with a
+`Pyramid  scaffold <https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/scaffolding.html>`_
+to create c2cgeoform-based applications. This scaffold produce a fully
+functional c2cgeoform-base project: the c2cgeoform_demo project.
+
+When running code checks and tests, these jobs are first run on the c2cgeoform
+framework itself where after the c2cgeoform_demo project is generated in `.build`
+folder and finally checks and tests are launched on this project too.
+
```
Comment (julsbreakdown) [inline docs/developer-guide.rst]: When running code checks and tests, these jobs are first run on the c2cgeoform
framework itself. Then the c2cgeoform_demo project is generated in `.build`
folder. Finally, the checks and tests are also launched in this project.

File: docs/developer-guide.rst (hunk 2 above)
Comment (julsbreakdown) [reply docs/developer-guide.rst]: Here is it => And there you go

Comment (julsbreakdown) [reply docs/developer-guide.rst]: you are ready to develop, make changes, run checks and tests in c2cgeoform. And finally see the results in ...

File: docs/developer-guide.rst (hunk 4 above)
Comment (julsbreakdown) [reply docs/developer-guide.rst]: Create the database as to serve the developement version?

File: docs/prerequisites.rst [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+-  ``python-virtualenv``
+-  ``libpq-dev`` (header files for PostgreSQL)
+-  ``gettext``
+
+On Windows, you should install ``make`` using Cygwin (and put the bin
+folder into the path). For Python, please use Python >= 3.x.
+
+You need to install PostgreSQL and PostGIS. On Ubuntu, the packages
+``postgresql-server-dev-9.3`` and ``python-dev`` are required.
```
Comment (julsbreakdown) [inline docs/prerequisites.rst]: postgres 9.3 => Minimum version?

File: docs/user-guide/grid.rst [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+Grid columns can be configured using the ``_list_fields`` property of the views
+class, which is an ordered list of ``ListField`` objects, one for each column.
+
+The ``ListField`` constructor take some parameters:
+
+* ``model``: the SQLAlchemy mapper (required if attr is an attribute name).
+* ``attr``: the model attribute name to use or an SQLAlchemy InstrumentedAttribute.
+* ``key``: an identifier for the column, default to ``attribute.key``.
+* ``label``: text for the column header, default to colanderalchemy title for the field.
+* ``renderer``: callable that take an entity of the SQLAlchemy mapper and
```
Comment (julsbreakdown) [inline docs/user-guide/grid.rst]: takes

Comment (julsbreakdown) [reply docs/user-guide/grid.rst]: returns

File: docs/user-guide/grid.rst [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  return a string value.
+* ``sort_column``: An ``IntrumentedAttribute`` to use in ``sort_by``.
+* ``filter_column``: An ``IntrumentedAttribute`` to filter with.
+* ``visible``: Is the column visible by default.
```
Comment (julsbreakdown) [inline docs/user-guide/grid.rst]: column will be visible by default

File: docs/user-guide/grid.rst [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+Each time the table index page will ask for data to your grid view, the
+``AbstractView`` will create a default query using ``AbstractViews._base_query`` method.
```
Comment (julsbreakdown) [inline docs/user-guide/grid.rst]: Every time the table index page asks for data from the grid view

File: docs/user-guide/grid.rst [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+
+If you use columns coming from relationships, this might result in sending one
+request to the database for each relationship and each record.
+In such cases, to avoid such performance issue you should override the
```
Comment (julsbreakdown) [inline docs/user-guide/grid.rst]: to avoid perfomance issues (no need of such)

File: docs/user-guide/model.rst [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+               'title': 'Name'
+           }})
+
+       comment = Column(Text, nullable=True, info={
+           'colanderalchemy': {
+               'title': 'Comment',
+               'widget': deform.widget.TextAreaWidget(rows=3),
+           }})
+
+This SQLAlchemy model is enriched with properties for
```
Comment (julsbreakdown) [inline docs/user-guide/model.rst]: This SQLAlchemy model is enriched with properties for
`ColanderAlchemy`_, for example to set a title for a field, use a
specific Deform `widget`_ or use a Colander `validator`_.

File: docs/user-guide/model.rst [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+`ColanderAlchemy`_, for example to set a title for a field, to use a
+specific Deform `widget`_ or to use a Colander `validator`_.
+
+In general, every SQLAlchemy model can be used as schema for a form. The
+only requirements are:
+
+-  The model class must contain exactly one primary key column. Tables
+   with composite primary keys are not supported.
+
+A more complex example for a model can be found `here`_. For more
```
Comment (julsbreakdown) [inline docs/user-guide/model.rst]: A more complex example for a model can be found `here`_. For more
information on how to define the model, please refer to the
`SQLAlchemy`_, `ColanderAlchemy`_, `Colander`_ and
`Deform`_ documentation.

File: docs/user-guide/project.rst [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+On Windows, you should use the https way to clone the repository:
+
+.. code-block:: shell
+
+   git clone https://github.com:camptocamp/c2cgeoform.git
+
+Create a Pyramid project using c2cgeoform scaffold
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Creating a ``c2cgeoform`` is creating a pyramid project using the
```
Comment (julsbreakdown) [inline docs/user-guide/project.rst]: Creating a ``c2cgeoform`` will create a

File: docs/user-guide/project.rst [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+c2cgeoform scaffold.
+
+Note that if PYTHONPATH does not exists as an environment variable,
```
Comment (julsbreakdown) [inline docs/user-guide/project.rst]: exists => exist

File: docs/user-guide/project.rst [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+template files (\*_tmpl) do not render in new project folder.
```
Comment (julsbreakdown) [inline docs/user-guide/project.rst]: do not render in new project folder. => are not rendered

File: docs/user-guide/project.rst [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+.. code-block:: shell
+
+   sudo -u postgres createdb c2cgeoform_project
+   sudo -u postgres psql -d c2cgeoform_project -c 'CREATE EXTENSION postgis;'
+   sudo -u postgres psql -c 'GRANT ALL ON DATABASE c2cgeoform_project TO "www-data";'
+
+When you do have a Postgres role and a PostGIS database edit the
+``development.ini`` and ``production.ini`` files and set
+``sqlachemy.url`` as appropriate. For example:
```
Comment (julsbreakdown) [inline docs/user-guide/project.rst]: ``sqlachemy.url`` appropriately. For example

File: docs/user-guide/project.rst [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+
+   sqlalchemy.url = postgresql://www-data:www-data@localhost:5432/c2cgeoform_project
+
+Now create the tables:
+
+.. code-block:: shell
+
+   make initdb
+
+Note that this execute the python script
```
Comment (julsbreakdown) [inline docs/user-guide/project.rst]: Note that this will execute

File: docs/user-guide/project.rst [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+``c2cgeoform_project/scripts/initializedb.py``. You will have to
+customize this thereafter.
+
+Run the development server
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You’re now ready to run the application:
```
Comment (julsbreakdown) [inline docs/user-guide/project.rst]: You are

File: docs/user-guide/schema.rst [hunk 26]
```diff
@@ -0,0 +1,23 @@
+Understand the schemas
```
Comment (julsbreakdown) [inline docs/user-guide/schema.rst]: Understanding

File: docs/user-guide/schema.rst [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+----------------------
+
+`ColanderAlchemy`_ allows creating `Colander`_ schemas directly from
+`SQLAlchemy`_ model classes.
+
+Additionally, c2cgeoform provides his own classes with extended features.
```
Comment (julsbreakdown) [inline docs/user-guide/schema.rst]: c2cgeoform provides its

File: docs/user-guide/schema.rst [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+In basic use case schema creation will look like:
```
Comment (julsbreakdown) [inline docs/user-guide/schema.rst]: A basic

File: docs/user-guide/schema.rst [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+
+.. code-block:: python
+
+   from model import MyClass
+   schema = GeoFormSchemaNode(MyClass)
+
+See following API to understand what is going on behind the scene.
```
Comment (julsbreakdown) [inline docs/user-guide/schema.rst]: See the following

File: docs/user-guide/templates.rst [hunk 30]
```diff
@@ -0,0 +1,77 @@
+.. _templates:
+
+Using custom templates
+----------------------
+
+c2cgeoform distinguishes between two types of templates: **views** templates
```
Comment (julsbreakdown) [inline docs/user-guide/templates.rst]: no need of between

File: docs/user-guide/templates.rst [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+
+.. _Jinja2: http://jinja.pocoo.org/
+
+Overriding widgets templates globally
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Deform widget templates are located in the ``templates/widgets`` folder and
+use the `chameleon`_ syntax.
+
+At rendering time, Deform will search for the templates as configured in Form
```
Comment (julsbreakdown) [inline docs/user-guide/templates.rst]: At rendering time, Deform will search for the templates that are configured in Form

File: docs/user-guide/templates.rst [hunk 32]
```diff
@@ -0,0 +1,77 @@
...
+.. code-block:: python
+
+   import c2cgeoform
+   search_paths = (
+       (resource_filename(__name__, 'templates/widgets'),) +
+       c2cgeoform.default_search_paths
+   )
+   c2cgeoform.default_search_paths = search_paths
+
+With this, to overwrite globally the `Deform templates`_ or the templates coming from
```
Comment (julsbreakdown) [inline docs/user-guide/templates.rst]: WIth this???  No need
Start the sentence with: To overwrite

File: docs/user-guide/templates.rst [hunk 33]
```diff
@@ -0,0 +1,77 @@
...
+``c2cgeoform`` (like the map widget), you just need to copy the template to your application
+``templates/widgets`` folder.
+
+.. _Chameleon: https://chameleon.readthedocs.org/en/latest/
+.. _Deform templates: https://github.com/Pylons/deform/tree/master/deform/templates
+
+Use a custom template for a form or a specific widget in a form
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The form main template as each widget template can be changed locally by for a
```
Comment (julsbreakdown) [inline docs/user-guide/templates.rst]: Both the form main template and widget templates can be c....

File: docs/user-guide/views.rst [hunk 34]
```diff
@@ -0,0 +1,64 @@
...
+Also note the ``@view_defaults`` which says that all the views declared in this
+class will only apply when the route parameter named ``table`` will be equal to
+``"excavation"``. The routes given by c2cgeoform have the following form:
+
+* c2cgeoform_index: ``{table}``
+* c2cgeoform_grid: ``{table}/grid.json``
+* c2cgeoform_item: ``{table}/{{id}}``
+* c2cgeoform_item_duplicate: ``{table}/{{id}}/duplicate``
+
+Those routes are registered in pyramid config with module ``routes``, see file
```
Comment (julsbreakdown) [inline docs/user-guide/views.rst]: in the pyramid config (see the routes.py file situated at the root of the..)

File: docs/user-guide/views.rst [hunk 35]
```diff
@@ -0,0 +1,64 @@
...
+   register_models(config, [
+       ('excavations', Excavation)
+
+To select records through urls, we also need a unique field, this is given by:
+
+.. code-block:: python
+
+   _id_field = 'hash'
+
+And to show the grid of table records we need a definition per column:
```
Comment (julsbreakdown) [inline docs/user-guide/views.rst]: And to show the table records grid...

File: docs/user-guide/widgets.rst [hunk 36]
```diff
@@ -0,0 +1,14 @@
+Configure the widgets
+---------------------
+
+All Deform widgets can be used with ``c2cgeoform``. See the Deform `examples`_
+and widgets `API reference`_ for detailed description about available options.
+
+Additionally, c2cgeoform provides some additionals widgets:
```
Comment (julsbreakdown) [inline docs/user-guide/widgets.rst]: additionals => extra

**Pull request merged**